File: Data Structures & Algorithms/sum-root-to-leaf-numbers/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        
        def solve(node, currVal):
            if not node:
                return 0
            currVal = currVal * 10 + node.val
            if not node.left and not node.right:
                return currVal
            return solve(node.left, currVal) + solve(node.right, currVal)
            
        return solve(root, 0)

File: Data Structures & Algorithms/balanced-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
private: 
    int solve(TreeNode* root, int &ans){
        if(!root) return 0;
        int lh = solve(root->left, ans);
        int rh = solve(root->right, ans);
        if(ans == -1) ans = -1;
        if(abs(lh-rh) > 1) ans = -1;
        return 1 + max(lh,rh);
    }
public:
    bool isBalanced(TreeNode* root) {
        int ans = 1;
        solve(root,ans);
        return ans == -1? false: true;
    }
};
